src/test/java/org/jenkinsci/plugins/workflow/flow/FlowExecutionListTest.java

```diff
@@ -160,6 +165,62 @@ public class FlowExecutionListTest {
         });
     }
 
+    @Ignore("Build never completes due to infinite loop")
+    @LocalData
+    @Test public void stepExecutionIteratorDoesNotLeakBuildsWhenOneIsStuck() throws Throwable {
+        // LocalData created using the following snippet while the build was waiting in the _second_ sleep, except
+        // for build.xml, which was captured during the sleep step. The StepEndNode for the stage was then adjusted to
+        // have its startId point to the timeout step's StepStartNode, creating a loop.
+        /*
+        sessions.then(r -> {
+            var stuck = r.createProject(WorkflowJob.class);
+            stuck.setDefinition(new CpsFlowDefinition("stage('stage') { sleep 30 }; timeout(time: 10) { sleep 30 }", true));
+            var b = stuck.scheduleBuild2(0).waitForStart();
+            System.out.println(b.getRootDir());
+            r.waitForCompletion(b);
+        });
+        */
+        sessions.then(r -> {
+            var p = r.jenkins.getItemByFullName("test0", WorkflowJob.class);
+            var b = p.getBuildByNumber(1);
+            // TODO: Create and run a new build, call StepExecutionIterator.applyAll while it is running, and assert
+            // that we do not leak a hard reference to it through the task queue for the CpsVmExecutorService for the
+            // stuck build and the futures inside of StepExecutionIteratorImpl.applyAll.
+            // TODO: Ideally, we would detect the issue with the stuck build and find some way to kill it.
+            r.waitForCompletion(b);
+        });
+    }
+
+    @Ignore("MemoryAssert is unable to clear the SoftReference from AbstractLazyLoadRunMap. Heap dumps show a strong reference from the test execution thread as a local variable, which is not captured by MemoryAssert. Also needs to be moved to a class with no BuildWatcher.")
```

Comment:
I was hoping to get this test to work since I'm not really sure if there is any other way to test the fix. It isn't totally clear to me why `MemoryAssert` isn't working. I think the strong reference from a local variable in the test thread that I saw might just be transient, for example I saw it referenced while the thread was at [this line](https://github.com/jenkinsci/jenkins-test-harness/blob/3585fb3db90fff5374405e4c1149863825147959/src/main/java/org/jvnet/hudson/test/MemoryAssert.java#L171). I am not sure though. I also tried configuring `BuildReference` to use weak references and saw the same behavior. Maybe I need to explicitly null out the local variable in the test.

Reply:
> Maybe I need to explicitly null out the local variable in the test.

🤦 that seems to fix it

Reply:
[b2c93a4](https://github.com/jenkinsci/workflow-api-plugin/pull/347/commits/b2c93a4ac1a2f8d957f440a4704924204c1ad8ef)